cpp/src/arrow/compute/kernels/select_k_test.cc

```diff
@@ -0,0 +1,737 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <functional>
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/array/array_decimal.h"
+#include "arrow/array/concatenate.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernels/test_util.h"
+#include "arrow/compute/kernels/util_internal.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_common.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/testing/util.h"
+#include "arrow/type_traits.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+namespace compute {
+
+template <typename ArrayType, SortOrder order>
+class SelectKComparator {
+ public:
+  template <typename Type>
+  bool operator()(const Type& lval, const Type& rval) {
+    if (order == SortOrder::Ascending) {
+      return lval <= rval;
+    } else {
+      return rval <= lval;
+    }
+  }
+};
+
+template <SortOrder order>
+Result<std::shared_ptr<Array>> SelectK(const Datum& values, int64_t k) {
+  if (order == SortOrder::Descending) {
+    return SelectKUnstable(values, SelectKOptions::TopKDefault(k));
+  } else {
+    return SelectKUnstable(values, SelectKOptions::BottomKDefault(k));
+  }
+}
+
+template <SortOrder order>
+Result<std::shared_ptr<Array>> SelectK(const Datum& values,
+                                       const SelectKOptions& options) {
+  if (order == SortOrder::Descending) {
+    return SelectKUnstable(Datum(values), options);
+  } else {
+    return SelectKUnstable(Datum(values), options);
+  }
+}
+
+void ValidateSelectK(const Datum& datum, int64_t k, Array& select_k_indices,
+                     SortOrder order, bool stable_sort = false) {
+  ASSERT_TRUE(datum.is_arraylike());
+  ASSERT_OK_AND_ASSIGN(auto sorted_indices,
+                       SortIndices(datum, SortOptions({SortKey("unused", order)})));
+
+  if (k < datum.length()) {
```

Comment:
Doesn't this mean we aren't asserting anything if k > datum.length()?

Reply:
this is a corner case, when k > datum.length() `k =  datum.length()` if there is not nulls. If there are nulls, then k is located at the beginning of these nulls after partition. So `k =  select_k_indices.length()`;